crates/ibc/src/clients/ics07_tendermint/client_state.rs

```diff
@@ -757,41 +823,29 @@ impl TryFrom<RawTmClientState> for ClientState {
             .frozen_height
             .and_then(|raw_height| raw_height.try_into().ok());
 
-        #[allow(deprecated)]
-        Ok(Self {
-            chain_id: ChainId::from_string(raw.chain_id.as_str()),
-            trust_level: trust_level
-                .try_into()
-                .map_err(|e| Error::invalid_trust_threshold(format!("{}", e)))?,
-            trusting_period: raw
-                .trusting_period
-                .ok_or_else(Error::missing_trusting_period)?
-                .try_into()
-                .map_err(|_| Error::negative_trusting_period())?,
-            unbonding_period: raw
-                .unbonding_period
-                .ok_or_else(Error::missing_unbonding_period)?
-                .try_into()
-                .map_err(|_| Error::negative_unbonding_period())?,
-            max_clock_drift: raw
-                .max_clock_drift
-                .ok_or_else(Error::missing_max_clock_drift)?
-                .try_into()
-                .map_err(|_| Error::negative_max_clock_drift())?,
-            latest_height: raw
-                .latest_height
-                .ok_or_else(Error::missing_latest_height)?
-                .try_into()
-                .map_err(|_| Error::missing_latest_height())?,
-            frozen_height,
-            upgrade_path: raw.upgrade_path,
-            allow_update: AllowUpdate {
-                after_expiry: raw.allow_update_after_expiry,
-                after_misbehaviour: raw.allow_update_after_misbehaviour,
-            },
-            proof_specs: raw.proof_specs.into(),
-            verifier: ProdVerifier::default(),
-        })
+        let allow_update = AllowUpdate {
+            after_expiry: raw.allow_update_after_expiry,
+            after_misbehaviour: raw.allow_update_after_misbehaviour,
+        };
+
+        let client_state = {
+            let mut cs = ClientState::new(
+                chain_id,
+                trust_level,
+                trusting_period,
+                unbonding_period,
+                max_clock_drift,
+                latest_height,
+                raw.proof_specs.into(),
+                raw.upgrade_path,
+                allow_update,
+            )?;
+
+            cs.frozen_height = frozen_height;
```

Comment:
why don't we modify the constructor to take in a `frozen_height`?

Reply:
I thought of `ClientState::new()` as a function that creates a new client and it wouldn't make sense to create a frozen client. This is also in accordance with the spec -> https://github.com/cosmos/ibc/blob/main/spec/client/ics-007-tendermint-client/README.md#client-initialisation
But I don't think we're checking for this in the create client hander (neither does ibc-go) -> https://github.com/cosmos/ibc-rs/issues/178

Reply:
> Therefore, I don't think we should use ClientState::new() here then

You're right :sweat_smile:. I used it here because we're doing all validation as part of `ClientState::new()` now. Would you prefer having a separate function `validate()` that is called by `ClientState::new()` and the `TryFrom` impl?

 > aren't we delegating that to `ctx.decode_client_state()` here?
 
 Yes, but I think this should be checked by the handler and we shouldn't rely on the host to give us an active client.

Reply:
> Would you prefer having a separate function validate() that is called by ClientState::new() and the TryFrom impl?

Well, do we even need `new()`? It is never used in the modules. If for some reason we prefer to keep it though, then sure a `validate()` would make sense.

> Yes, but I think this should be checked by the handler and we shouldn't rely on the host to give us an active client.

Agreed.

Reply:
I decided to keep `new()` and add a `frozen_height` parameter to it as initially suggested. I like to keep the `new()` method as it provides a way to create a correct `ClientState` (without requiring a raw struct), and it's also being used in tests (otherwise we'd need to make the fields `pub(crate)`).

Reply:
https://github.com/cosmos/ibc-rs/pull/172/commits/27f3cbf5da92920c3e3ff25431de765219a05288 and https://github.com/cosmos/ibc-rs/pull/172/commits/355a4a152189ba5dd304dad714009abec155ff89